Extension/src/LanguageServer/lmTool.ts

```diff
@@ -44,6 +48,129 @@ const knownValues: { [Property in keyof ChatContextResult]?: { [id: string]: str
+    const filterMap = getCompilerArgumentFilterMap(compiler, context);
+    if (filterMap === undefined) {
+        return [];
+    }
+
+    const filteredArguments: string[] = [];
+    for (const key in filterMap) {
+        if (!key) {
+            continue;
+        }
+        const filter = new RegExp(key);
+        const filtered = compilerArguments.filter(arg => filter.test(arg));
+        filteredArguments.push(...filtered);
+    }
+    return filteredArguments;
+}
+
+export async function getProjectContext(uri: vscode.Uri, context: { flags: Record<string, unknown> }, token: vscode.CancellationToken): Promise<ProjectContext | undefined> {
+    const telemetryProperties: Record<string, string> = {};
+    try {
+        const projectContext = await checkTime<ProjectContextResult | undefined>(async () => await getClients()?.ActiveClient?.getProjectContext(uri, token) ?? undefined);
+        telemetryProperties["time"] = projectContext.time.toFixed(1);
+        if (!projectContext.result) {
+            return undefined;
+        }
+
+        formatChatContext(projectContext.result);
+
+        const result: ProjectContext = {
+            language: projectContext.result.language,
+            standardVersion: projectContext.result.standardVersion,
+            compiler: projectContext.result.compiler,
+            targetPlatform: projectContext.result.targetPlatform,
+            targetArchitecture: projectContext.result.targetArchitecture,
+            compilerArguments: []
+        };
+
+        if (projectContext.result.language) {
+            telemetryProperties["language"] = projectContext.result.language;
+        }
+        if (projectContext.result.compiler) {
+            telemetryProperties["compiler"] = projectContext.result.compiler;
+        }
+        if (projectContext.result.standardVersion) {
+            telemetryProperties["standardVersion"] = projectContext.result.standardVersion;
+        }
+        if (projectContext.result.targetPlatform) {
+            telemetryProperties["targetPlatform"] = projectContext.result.targetPlatform;
+        }
+        if (projectContext.result.targetArchitecture) {
+            telemetryProperties["targetArchitecture"] = projectContext.result.targetArchitecture;
+        }
+        telemetryProperties["compilerArgumentCount"] = projectContext.result.fileContext.compilerArguments.length.toString();
+        // Telemetry to learn about the argument distribution. The filtered arguments are expected to be non-PII.
+        if (projectContext.result.fileContext.compilerArguments.length) {
+            const filteredCompilerArguments = filterComplierArguments(projectContext.result.compiler, projectContext.result.fileContext.compilerArguments, context);
+            if (filteredCompilerArguments.length > 0) {
+                telemetryProperties["filteredCompilerArguments"] = filteredCompilerArguments.join(' ');

```

Comment:
Can this telemetry be moved up, after the duplicates are removed?

Reply:
Up to the code that also calls .filter but handles the directAsksMap as well. It seems like that processing should just be done in one location, unless I'm not understanding something.

Reply:
Fixed the code to only filter once, thanks.